// notiontheory-basic-build/src/setup.js
var bump = require("gulp-bump"),
  exec = require("./exec"),
  fs = require("fs"),
  path = require("path");

function setup(gulp, pkg) {

  pkg = pkg && JSON.parse(JSON.stringify(pkg)) || {};
  pkg.packageName = pkg.name;

  var leBump = function(type){
    return function(){
      return gulp.src("./package.json")
        .pipe(bump({ type: type }))
        .pipe(gulp.dest("./"));
    };
  };

  gulp.task("bump", leBump("patch"));
  gulp.task("bumpMinor", leBump("minor"));
  gulp.task("bumpMajor", leBump("major"));
  gulp.task("yolo", function(cb) {
    fs.readFile(path.join(process.cwd(), "package.json"), function(err, file) {
      if(err){
        cb(err);
      }
      else{
      var info = JSON.parse(file),
        git = "gulp release && git add -A . && git commit -m \"v" + info.version + "\" && git push";
        exec(git, cb);
      }
    });
  });

  gulp.task("trololo", exec("git checkout master && git pull && git merge dev && git push && git checkout dev"));

  return {
    js: require("./js").bind(this, gulp, pkg),
    cat: require("./cat").bind(this, gulp, pkg),
    min: require("./min").bind(this, gulp, pkg),
    html: require("./html").bind(this, gulp, pkg),
    css: require("./css").bind(this, gulp, pkg),
    clean: require("./clean").bind(this, gulp, pkg)
  };
}

module.exports = setup;
